finetuner-workflow/finetuner/finetuner.py

```diff
@@ -266,93 +645,102 @@ def __getitem__(self, idx) -> Tuple[torch.Tensor, torch.Tensor]:
 
 if args.no_shuffle:
-     train_dataset = dataset
-     val_dataset = RandomSampler(dataset, num_samples=len(dataset) - train_size)
+    train_dataset = dataset
+    val_dataset = RandomSampler(dataset, num_samples=len(dataset) - train_size)
 else:
-     train_dataset, val_dataset = random_split(dataset,
-                                               [train_size,
-                                                len(dataset) - train_size])
-print(f"TRAIN_DATASET: {len(train_dataset):,} examples")
-print(f"VALUE_DATASET: {len(val_dataset):,} examples")
-
-# Where we write our training checkpoints and final model.
-output_dir = os.path.abspath(
-    os.path.join(args.output_path, "results-" + args.run_name))
-
-
-# Properly type-cast the param (str to bool)
-FALSE = [
-    "False",
-    "false",
-    "f",
-    "0",
-]
-args.no_resume = args.no_resume not in FALSE
-
-# Discover if we have any checkpoints to resume from.
-if not args.no_resume:
-    try:
-        output_dir_list = os.listdir(output_dir)
-        lastCheckpoint = sorted(output_dir_list,
-                                key=lambda x: int(x.split("-")[1]))[-1]
-    except:
-        lastCheckpoint = None
-else:
-    lastCheckpoint = None
-print("LAST CHECKPOINT:", lastCheckpoint)
-
-# Set random seed, for ML research purposes and reproducibility, it's important
-# that we set this to a consistent value.
-torch.manual_seed(args.seed)
-print("RANDOM SEED:", args.seed)
+    train_dataset, val_dataset = random_split(
+        dataset, [train_size, len(dataset) - train_size]
+    )
 
 # Determine if we train in fp32 or fp16 mode.
-print("FORCE FP16:", args.fp16)
-fp16_arg = {}
+trainer_fp16_args = {}
 if args.fp16:
-    fp16_arg = {'fp16': True}
+    trainer_fp16_args["fp16"] = True
+if args.fp16_full_eval:
+    trainer_fp16_args["fp16_full_eval"] = True
+
+if is_main_process():
+    logger.info(f"TRAIN_DATASET: {len(train_dataset):,} examples")
+    logger.info(f"VALUE_DATASET: {len(val_dataset):,} examples")
+    logger.info(f"RANDOM SEED: {args.seed}")
+
 
 # Load our model that we're training. This may fetch via HTTP if not cached
 # already.
-print(f"Loading {args.model}")
-if 'tensorizer' in sys.modules and validators.url(args.model):
-    tokenizer, unitrim, word_tokens = get_tokenizer(args.model,
-                                                    {"eos_token": args.eot,
-                                                     "pad_token": args.pad})
-    if args.fp16:
-        model = load_model(args.model).half()
-    else:
-        model = load_model(args.model)
+model: PreTrainedModel
 
-else:
-    try:
-        tokenizer = AutoTokenizer.from_pretrained(args.model,
-                                                  eos_token=args.eot,
-                                                  pad_token=args.pad,
-                                                  cache_dir=args.cache)
+try:
+    with no_init():
+        model_fp16_args = {"torch_dtype": torch.float16}
```

Comment:
`model_fp16_args` here looks like it was intended to only be passed conditionally in `FORCE_FP16` mode, but it's passed either way at the moment. Although, I actually [removed the downcast to fp16 here in an earlier commit](https://github.com/coreweave/kubernetes-cloud/commit/78e4a10243a483f5b162d4d573b268bfae2843dc) (when it was using `.half()`) even for the fp16 (actually mixed precision) mode after some testing showed that:
- It didn't make training or evaluation faster,
- It didn't decrease VRAM usage, and
- It's a completely invalid configuration with base the `transformers` library and only works because `deepspeed` has more lenient runtime checks

I think this was because DeepSpeed's mixed precision training casts most of the stored model weights back up to fp32 anyway, or something like that, so you lose precision with no particular benefit.
If casting it like this works substantially differently than calling `.half()` did before, though, somehow, maybe it could be helpful.

Reply:
#186 fixes this to respect `FORCE_FP16` / `--fp16`, but the concern about whether or not to downcast the model at all at this step could still use some investigation.